torch/csrc/jit/python/python_ivalue.h

```diff
@@ -24,6 +35,11 @@ struct C10_EXPORT ConcretePyObjectHolder final : PyObjectHolder {
     return py_obj_.ptr();
   }
 
+  IValue toIValue() override {
```

Comment:
If you have suggestions on how to make it valid for all cases I'd love to hear it! That's exactly what I'm trying to achieve here! For now, indeed, this only works for Python objects whose type can be inferred.

If you this it helps I can rename this method to `toTypeInferredIValue`, to clarify that this behaves like the global function `torch::jit::toTypeInferredIValue`, which also raises if it fails. WDYT?

Reply:
Alternatively I guess I could make this return an `optional<IValue>` (I would have to reimplement some of the logic of `torch::jit::toTypeInferredIValue`). Would that work?

Reply:
can we add two apis here for PyObjectHolder? 

1. `bool isIValue()`: returns true only when we could infer a type by calling `tryToInferType` and `match.success()`. 
2. `toIValue()`: which does what you currently do

Then the callsites in `getSubValues` should be guarded by `isIValue()`, otherwise fallback to old behaviors. what do you think about this solution?

Reply:
This makes sense to me, and I did something very similar to what you just suggested: instead of `isIValue` I added a `tryToInferType` method, which returns the TypePtr (or the reason this failed). I opted to go that way in order to keep the alignment between the global functions of `torch::jit` and the methods of `ConcretePyObjectHolder` (so that it's easy to see that they do the same thing).